File: Algoexpert/Validate BST/java/src/ValidateBST.java
public class ValidateBST {

    public static boolean validateBst(BST tree) {
        return validateBstHelper(tree, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    public static boolean validateBstHelper(BST tree, int minValue, int maxValue) {

        if (tree.value <  minValue || tree.value >= maxValue) {
            return false;
        }

        if (tree.left != null && !validateBstHelper(tree.left, minValue, tree.value)) {
            return false;
        }

        if (tree.right != null && !validateBstHelper(tree.right, tree.value, maxValue)) {
            return false;
        }

        return true;
    }

    public static void main(String[] args) {

        BST root = new BST(10);
        root.left = new BST(5);
        root.left.left = new BST(2);
        root.left.left.left = new BST(1);
        root.left.right = new BST(5);
        root.right = new BST(15);
        root.right.left = new BST(13);
        root.right.left.right = new BST(14);
        root.right.right = new BST(22);

        System.out.println(validateBst(root));
    }

}

File: Algoexpert/Validate BST/cpp/validate_bst.cpp
#include <bits/stdc++.h>

using namespace std;

class BST {
    public:
        int value;
        BST *left;
        BST *right;

        BST(int val) {
            value = val;
            left = nullptr;
            right = nullptr;
        }
};

class Solution {
    public:
        bool validateBst(BST *tree) {
            return validateBSTHelper(tree, INT_MIN, INT_MAX);
        }

        bool validateBSTHelper(BST *tree, int minValue, int maxValue) {
            
            if (tree->value < minValue || tree->value >= maxValue) {
                return false;
            }
            
            if(tree->left != nullptr && !validateBSTHelper(tree->left, minValue, tree->value)) {
                return false;
            }
                
            if(tree->right != nullptr && !validateBSTHelper(tree->right, tree->value, maxValue)) {
                return false;
            }
            
            return true;
        }
};


int main() {

    BST *root = new BST(10);
    root->left = new BST(5);
    root->left->left = new BST(2);
    root->left->left->left = new BST(1);
    root->left->right = new BST(5);
    root->right = new BST(15);
    root->right->left = new BST(13);
    root->right->left->right = new BST(14);
    root->right->right = new BST(22);

    Solution solution;
    cout <<  solution.validateBst(root) << endl;

    return 0;
}